File: server/token_server/ruby/lib/token_server.rb
require 'base64'
require 'json'
require 'digest/sha1'
require 'time'
require 'zlib'
require 'openssl'

class TokenServer
  attr_accessor :app_key, :app_secret, :default_ttl_sec

  def initialize(app_key, app_secret, default_ttl_sec)
    raise ArgumentError, "appKey or appSecret is empty" if app_key.empty? || app_secret.empty?
    raise ArgumentError, "defaultTTLSec must be positive" if default_ttl_sec <= 0

    @app_key = app_key
    @app_secret = app_secret
    @default_ttl_sec = default_ttl_sec
  end

  def get_token(channel_name, uid, ttl_sec = nil)
    ttl_sec = @default_ttl_sec if ttl_sec.nil? || ttl_sec <= 0
    cur_time = (Time.now.to_f * 1000).to_i

    sign_str = "#{@app_key}#{uid}#{cur_time}#{ttl_sec}#{channel_name}#{@app_secret}"
    signature = Digest::SHA1.hexdigest(sign_str)

    token_model = {
      signature: signature,
      curTime: cur_time,
      ttl: ttl_sec
    }

    Base64.strict_encode64(token_model.to_json)
  end

  def get_permission_key(channel_name, perm_secret, uid, privilege, ttl_sec)
    cur_time = Time.now.to_i
    get_permission_key_with_current_time(channel_name, perm_secret, uid, privilege, ttl_sec, cur_time)
  end

  private

  # 使用指定时间生成权限密钥
  def get_permission_key_with_current_time(channel_name, perm_secret, uid, privilege, ttl_sec, cur_time)
    # 计算校验和
    checksum = hmac_sha256(
      @app_key,
      uid.to_s,
      cur_time.to_s,
      ttl_sec.to_s,
      channel_name,
      perm_secret,
      privilege.to_s
    )

    # 构建权限键映射
    perm_key_map = {
      "appkey" => @app_key,
      "checksum" => checksum,
      "cname" => channel_name,
      "curTime" => cur_time,
      "expireTime" => ttl_sec,
      "privilege" => privilege,
      "uid" => uid,
    }

    json_data = perm_key_map.to_json
    compressed_data = Zlib.deflate(json_data)
    custom_base64_encode(compressed_data)
  end

  # 计算 HMAC-SHA256 签名
  def hmac_sha256(appid_str, uid_str, cur_time_str, expire_time_str, cname, perm_secret, privilege_str)
    content = [
      "appkey:#{appid_str}",
      "uid:#{uid_str}",
      "curTime:#{cur_time_str}",
      "expireTime:#{expire_time_str}",
      "cname:#{cname}",
      "privilege:#{privilege_str}",
      ""  # 最后一行换行符
    ].join("\n")

    digest = OpenSSL::HMAC.digest('sha256', perm_secret, content)
    Base64.strict_encode64(digest)
  end

  # 自定义 Base64 编码
  def custom_base64_encode(data)
    # 标准 Base64 编码
    standard_encoded = Base64.strict_encode64(data)

    # 替换字符集和填充符
    standard_encoded
      .tr('+/', '*-')   # 替换 + 为 *，/ 为 -
      .gsub('=', '_')   # 替换 = 为 _
  end
end

File: android/Advance/aicapacity/src/main/java/com/netease/nertc/AICapacity/TokenServer.java
package com.netease.nertc.aicapacity;

import android.os.Build;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.zip.Deflater;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.security.InvalidKeyException;

public class TokenServer {
    private final String appKey;
    private final String appSecret;
    private final int defaultTTLSec;

    public TokenServer(String appKey, String appSecret, int defaultTTLSec) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            if (appKey == null || appSecret == null || appKey.isEmpty() || appSecret.isEmpty()) {
                throw new IllegalArgumentException("appKey or appSecret is empty");
            }
        }
        if (defaultTTLSec <= 0) {
            throw new IllegalArgumentException("defaultTTLSec must be positive");
        }
        this.appKey = appKey;
        this.appSecret = appSecret;
        this.defaultTTLSec = defaultTTLSec;
    }

    /**
     * Generate a token with the given channel name, user ID, and time-to-live (TTL) in seconds.
     *
     * @param channelName: the name of the channel for which the token is being generated
     * @param uid:         the user ID associated with the token
     * @param ttlSec:      the time-to-live (TTL) for the token in seconds, should be less than 86400 seconds (1 day)
     * @return: the token
     */
    public String getToken(String channelName, long uid, int ttlSec) throws Exception {
        long curTimeMs = System.currentTimeMillis();
        return getTokenWithCurrentTime(channelName, uid, ttlSec, curTimeMs);
    }

    public String getTokenWithCurrentTime(String channelName, long uid, int ttlSec, long curTimeMs) throws Exception {
        if (ttlSec <= 0) {
            ttlSec = defaultTTLSec;
        }

        // 计算签名
        String rawSignature = sha1(String.format("%s%d%d%d%s%s", appKey, uid, curTimeMs, ttlSec, channelName, appSecret));

        // 使用JSONObject替代ObjectMapper
        JSONObject tokenModel = new JSONObject();
        tokenModel.put("signature", rawSignature);
        tokenModel.put("curTime", curTimeMs);
        tokenModel.put("ttl", ttlSec);

        // 生成JSON字符串
        String signature = tokenModel.toString();

        // 进行Base64编码
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            return Base64.getEncoder().encodeToString(signature.getBytes(StandardCharsets.UTF_8));
        }

        // 对于API级别低于26的情况，可以添加兼容性处理
        return "";
    }

    /**
     * generates a permission key for the given channelName, uid and privilege.
     *
     * @param channelName: the name of the channel for which the token is being generated
     * @param permSecret:  the secret key for generating the permission key
     * @param uid:         the user ID associated with the token
     * @param privilege:   the privilege of the user. privilege is a 8-bit number, each bit represents a permission:
     *                     - bit 1: (0000 0001) = 1, permission to send audio stream
     *                     - bit 2: (0000 0010) = 2, permission to send video stream
     *                     - bit 3: (0000 0100) = 4, permission to receive audio stream
     *                     - bit 4: (0000 1000) = 8, permission to receive video stream
     *                     - bit 5: (0001 0000) = 16, permission to create room
     *                     - bit 6: (0010 0000) = 32, permission to join room
     *                     So, (0010 1100) = 32+8+4 = 44 means permission to receive audio&video stream and join room.
     *                     (0011 1111) = 63 means all permission allowed.
     * @param ttlSec:      the time-to-live (TTL) for the token in seconds, should be less than 86400 seconds (1 day)
     * @return the permission key
     * @throws NoSuchAlgorithmException
     * @throws InvalidKeyException
     * @throws JsonProcessingException
     */
    public String getPermissionKey(String channelName, String permSecret, long uid, byte privilege, long ttlSec)
            throws NoSuchAlgorithmException, InvalidKeyException {
        long curTime = System.currentTimeMillis() / 1000;
        try {
            return getPermissionKeyWithCurrentTime(channelName, permSecret, uid, privilege, ttlSec, curTime);
        } catch (RuntimeException e) {
            // 处理JSON构建异常
            Throwable cause = e.getCause();
            if (cause instanceof JSONException) {
                throw new RuntimeException("生成权限密钥失败: " + cause.getMessage(), cause);
            }
            throw e;
        }
    }

    public String getPermissionKeyWithCurrentTime(String channelName, String permSecret, long uid, byte privilege,
                                                  long ttlSec, long curTime) throws NoSuchAlgorithmException, InvalidKeyException {
        // 构建权限密钥对象
        PermissionKey permKey = new PermissionKey();
        permKey.appkey = appKey;
        permKey.uid = uid;
        permKey.cname = channelName;
        permKey.privilege = privilege;
        permKey.expireTime = ttlSec;
        permKey.curTime = curTime;

        // 计算签名
        String checksum = hmacsha256(appKey, String.valueOf(uid), String.valueOf(curTime), String.valueOf(ttlSec),
                channelName, permSecret, String.valueOf(privilege));
        permKey.checksum = checksum;

        // 使用JSONObject替代ObjectMapper
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("appkey", permKey.appkey);
            jsonObject.put("uid", permKey.uid);
            jsonObject.put("cname", permKey.cname);
            jsonObject.put("privilege", permKey.privilege);
            jsonObject.put("expireTime", permKey.expireTime);
            jsonObject.put("curTime", permKey.curTime);
            jsonObject.put("checksum", permKey.checksum);

            Map<String, Object> sortedMap = new TreeMap<>();
            Iterator<String> keys = jsonObject.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                sortedMap.put(key, jsonObject.get(key));
            }

            String jsonStr = new JSONObject(sortedMap).toString();

            // 压缩数据
            byte[] compressedData = null;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                compressedData = compress(jsonStr.getBytes(StandardCharsets.UTF_8));
            }

            // 编码压缩后的数据并返回
            return new String(base64EncodeUrl(compressedData));

        } catch (JSONException e) {
            throw new RuntimeException("构建JSON字符串失败", e);
        }
    }

    private static byte[] compress(byte[] data) {
        try {
            Deflater deflater = new Deflater(6);
            deflater.setInput(data);
            deflater.finish();

            byte[] buffer = new byte[2048];
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream(data.length);
            while (!deflater.finished()) {
                int count = deflater.deflate(buffer);
                outputStream.write(buffer, 0, count);
            }
            outputStream.close();

            byte[] compressedData = outputStream.toByteArray();
            deflater.end();
            return compressedData;
        } catch (java.io.IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static String hmacsha256(String appidStr, String uidStr, String curTimeStr, String expireTimeStr,
                                     String cname, String permSecret, String privilegeStr)
            throws NoSuchAlgorithmException, InvalidKeyException {
        String contentToBeSigned = "appkey:" + appidStr + "\n";
        contentToBeSigned += "uid:" + uidStr + "\n";
        contentToBeSigned += "curTime:" + curTimeStr + "\n";
        contentToBeSigned += "expireTime:" + expireTimeStr + "\n";
        contentToBeSigned += "cname:" + cname + "\n";
        contentToBeSigned += "privilege:" + privilegeStr + "\n";

        SecretKeySpec keySpec = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                keySpec = new SecretKeySpec(permSecret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            }
        }
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(keySpec);
        byte[] result = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            result = mac.doFinal(contentToBeSigned.getBytes(StandardCharsets.UTF_8));
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                return Base64.getEncoder().encodeToString(result);
            }
        }
        return "";
    }

    private String sha1(String input) throws NoSuchAlgorithmException {
        MessageDigest mDigest = MessageDigest.getInstance("SHA-1");
        byte[] result = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                result = mDigest.digest(input.getBytes(StandardCharsets.UTF_8));
            }
        }
        StringBuilder sb = new StringBuilder();
        for (byte b : result) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    public static byte[] base64EncodeUrl(byte[] input) {
        byte[] base64 = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            base64 = Base64.getEncoder().encode(input);
        }
        for (int i = 0; i < base64.length; ++i)
            switch (base64[i]) {
                case '+':
                    base64[i] = '*';
                    break;
                case '/':
                    base64[i] = '-';
                    break;
                case '=':
                    base64[i] = '_';
                    break;
                default:
                    break;
            }
        return base64;
    }

    public static class DynamicToken {
        public String signature;
        public long curTime;
        public int ttl;
    }

    public static class PermissionKey {
        public String appkey;
        public long uid;
        public String cname;
        public byte privilege;
        public long expireTime;
        public long curTime;
        public String checksum;
    }
}
